File: frontend/src/lib/services/productionService.js
const API_URL = import.meta.env.VITE_API_URL || '';

function getHeaders(extraHeaders = {}) {
    const role = localStorage.getItem('user_role') || 'admin';
    return {
        'X-User-Role': role,
        ...extraHeaders
    };
}

export class ProductionService {
    constructor(api_url) {
        this.baseUrl = `${api_url}/jobs`;
    }

    async fetchJobs(skip = 0, limit = 100, filters = {}) {
        const params = new URLSearchParams({ skip: skip.toString(), limit: limit.toString(), ...filters });
        const response = await fetch(`${this.baseUrl}/?${params.toString()}`, {
            headers: getHeaders()
        });
        if (!response.ok) throw new Error('Failed to fetch jobs');
        return await response.json();
    }

    async getJob(id) {
        const response = await fetch(`${this.baseUrl}/${id}`, {
            headers: getHeaders()
        });
        if (!response.ok) throw new Error('Failed to fetch job');
        return await response.json();
    }

    async createJob(job) {
        const response = await fetch(`${this.baseUrl}/`, {
            method: 'POST',
            headers: getHeaders({ 'Content-Type': 'application/json' }),
            body: JSON.stringify(job),
        });
        if (!response.ok) throw new Error('Failed to create job');
        return await response.json();
    }

    async updateJob(id, job) {
        const response = await fetch(`${this.baseUrl}/${id}`, {
            method: 'PUT',
            headers: getHeaders({ 'Content-Type': 'application/json' }),
            body: JSON.stringify(job),
        });
        if (!response.ok) throw new Error('Failed to update job');
        return await response.json();
    }

    async updateJobStatus(id, status) {
        const response = await fetch(`${this.baseUrl}/${id}/status?status=${encodeURIComponent(status)}`, {
            method: 'PATCH',
            headers: getHeaders(),
        });
        if (!response.ok) throw new Error('Failed to update job status');
        return await response.json();
    }

    async deleteJob(id) {
        const response = await fetch(`${this.baseUrl}/${id}`, {
            method: 'DELETE',
            headers: getHeaders()
        });
        if (!response.ok) throw new Error('Failed to delete job');
        return await response.json();
    }

    async fetchOrderTasks(orderId) {
        const response = await fetch(`${API_URL}/gnc/orders/${orderId}/tasks`, {
            headers: getHeaders(),
        });
        if (!response.ok) throw new Error('Failed to fetch order tasks');
        return await response.json();
    }

    async saveOrderNesting(project) {
        const response = await fetch(`${API_URL}/gnc/save-order-nesting`, {
            method: 'POST',
            headers: getHeaders({ 'Content-Type': 'application/json' }),
            body: JSON.stringify(project),
        });
        if (!response.ok) throw new Error('Failed to save order nesting');
        return await response.json();
    }

    async fetchOrderNestingProject(orderId) {
        const response = await fetch(`${API_URL}/gnc/orders/${orderId}/project`, {
            headers: getHeaders(),
        });
        if (!response.ok) throw new Error('Failed to fetch nesting project');
        return await response.json();
    }

    async saveAsNewOrder(name, sheets, originalDocumentId) {
        const response = await fetch(`${API_URL}/documents/save-as-new-order`, {
            method: 'POST',
            headers: getHeaders({ 'Content-Type': 'application/json' }),
            body: JSON.stringify({ name, sheets, original_document_id: originalDocumentId }),
        });
        if (!response.ok) throw new Error('Failed to save as new order');
        return await response.json();
    }
}

File: old mvp/backend/src/application/services/production_service.py
import json

from src.domain.interfaces import IDocumentRepository, ITaskRepository
from src.domain.models import Document, Task


class ProductionService:
    def __init__(self, task_repo: ITaskRepository, doc_repo: IDocumentRepository):
        self.task_repo = task_repo
        self.doc_repo = doc_repo

    def list_jobs(self, skip: int = 0, limit: int = 100, filters: dict = None) -> list[Task]:
        return self.task_repo.list(skip, limit, filters)

    def get_job(self, id: int) -> Task | None:
        return self.task_repo.get_by_id(id)

    def create_job(self, task: Task) -> Task:
        return self.task_repo.add(task)

    def update_job(self, task: Task) -> Task:
        return self.task_repo.update(task)

    def delete_job(self, id: int) -> bool:
        return self.task_repo.delete(id)

    def update_job_status(self, id: int, status: str) -> Task | None:
        task = self.task_repo.get_by_id(id)
        if task:
            task.status = status
            return self.task_repo.update(task)
        return None

    def get_order_tasks(self, order_id: int) -> list[Task]:
        return self.task_repo.get_tasks_by_document_id(order_id)

    def save_nesting(self, order_id: int, project_data: dict) -> bool:
        doc = self.doc_repo.get_by_id(order_id)
        if doc:
            doc.content = json.dumps(project_data)
            self.doc_repo.update(doc)
            return True
        return False

    def get_nesting_project(self, order_id: int) -> dict | None:
        doc = self.doc_repo.get_by_id(order_id)
        if doc and doc.content:
            try:
                return json.loads(doc.content)
            except:
                return None
        return None

    def save_as_new_order(self, data: dict) -> Document:
        return self.doc_repo.save_as_new_order(data)
